lib/ccf/index.ts

```diff
@@ -0,0 +1,259 @@
+        this.name = name;
+        if (staticParams === undefined) {
+            throw new Error('Required Parameters not provided');
+        }
+        if ('provider' in staticParams) {
+            const provider = staticParams?.provider as string;
+            if (['aws', 'gcp', 'azure'].includes(provider)) {
+                this.provider = provider;
+            } else {
+                throw new Error('Provider not supported');
+            }
+        }
+        if ('instance_type' in staticParams) {
+            const instanceType = staticParams?.instance_type as string;
+            if (instanceType in this.computeInstances[this.provider]) {
+                this.instanceType = instanceType;
+            } else {
+                throw new Error('Instance Type not supported');
+            }
+        }
+        if ('expected_lifespan' in staticParams) {
+            this.expectedLifespan = staticParams?.expected_lifespan as number;
+        }
+        return this;
+    }
+
+    async calculate(observations: object | object[] | undefined): Promise<object> {
+        if (observations === undefined) {
+            throw new Error('Required Parameters not provided');
+        }
+        if (this.instanceType === '' || this.provider === '') {
+            throw new Error('Configuration is incomplete');
+        }
+        // let mTotal = this.embodiedEmissions();
+        const results: any[] = [];
+        if (Array.isArray(observations)) {
+            observations.forEach((observation: { [key: string]: any }) => {
+                const e = this.calculateEnergy(observation);
+                const m = this.embodiedEmissions(observation);
+                results.push({
+                    "e": e,
+                    "m": m,
+                });
+            });
+
+        }
+        return results;
+    }
+
+    calculateEnergy(observation: { [key: string]: any; }) {
+        if (!('duration' in observation) || !('cpu' in observation) || !('datetime' in observation)) {
+            throw new Error('Required Parameters duration,cpu,datetime not provided for observation');
+        }
+        //    duration is in seconds
+        const duration = observation['duration'];
+        //    convert cpu usage to percentage
+        const cpu = observation['cpu'] * 100.0;
+        //  get the wattage for the instance type
+        let wattage = 0;
+        if (this.provider === 'aws') {
+            const x = [0, 10, 50, 100];
+            const y: number[] = [
+                this.computeInstances['aws'][this.instanceType].consumption.idle ?? 0,
+                this.computeInstances['aws'][this.instanceType].consumption.tenPercent ?? 0,
+                this.computeInstances['aws'][this.instanceType].consumption.fiftyPercent ?? 0,
+                this.computeInstances['aws'][this.instanceType].consumption.hundredPercent ?? 0
+            ];
+            const spline = new Spline(x, y);
```

Comment:
We should support different methods here. 
- Spline is one.
- Linear interpolation between points is another.
- Might have others?

Reply:
Linear Interpolation is used for GCP & Azure via the Architecture Type. 

As suggested by Cameron, I am now planning to use the hand-written list of instances from the CCF Repository for determining the architecture of an instance. 

We will lose the Spline interpolation for AWS once such is implemented. 

I still do carry the question whether the this [aws instances use](https://github.com/cloud-carbon-footprint/ccf-coefficients/blob/main/output/coefficients-aws-use.csv) data is for total instance usage or just for CPU ? Why there is a GB/Chip metric if not?

Reply:
I have now implemented linear interpolation using the handwritten definition in the ccf repository.